src/Parser.zig

```diff
@@ -3635,50 +3638,219 @@ fn convertInitList(p: *Parser, il: InitList, init_ty: Type) Error!NodeIndex {
     }
 }
 
+fn parseMSVCAsmStmt(p: *Parser) Error!?NodeIndex {
+    return p.todo("MSVC assembly statements");
+}
+
+fn parseAsmOperands(p: *Parser, names: *std.ArrayList(?TokenIndex), constraints: *NodeList, exprs: *NodeList) Error!void {
+    if (!p.tok_ids[p.tok_i].isStringLiteral() and p.tok_ids[p.tok_i] != .l_bracket) {
+        // Empty
+        return;
+    }
+    while (true) {
+        if (p.eatToken(.l_bracket)) |l_bracket| {
+            const ident = (try p.eatIdentifier()) orelse {
+                try p.err(.expected_identifier);
+                return error.ParsingFailed;
+            };
+            try names.append(ident);
+            try p.expectClosing(l_bracket, .r_bracket);
+        } else {
+            try names.append(null);
+        }
+        const constraint = try p.asmStr();
+        try constraints.append(constraint.node);
+
+        const l_paren = p.eatToken(.l_paren) orelse {
+            try p.errExtra(.expected_token, p.tok_i, .{ .tok_id = .{ .actual = p.tok_ids[p.tok_i], .expected = .l_paren } });
+            return error.ParsingFailed;
+        };
+        const res = try p.expr();
+        try p.expectClosing(l_paren, .r_paren);
+        try res.expect(p);
+        try exprs.append(res.node);
+        if (p.eatToken(.comma) == null) return;
+    }
+}
+fn parseGNUAsmStmt(p: *Parser, asm_str_node: NodeIndex, quals: Tree.GNUAssemblyQualifiers) Error!NodeIndex {
+    if (p.tok_ids[p.tok_i] == .r_paren) {
+        return p.addNode(.{
+            .tag = .gnu_asm_simple,
+            .ty = .{ .specifier = .void },
+            .data = .{ .un = asm_str_node },
+        });
+    }
+
+    const expected_items = 8; // arbitrarily chosen, most assembly will have fewer than 8 inputs/outputs/constraints/names
+    const bytes_needed = expected_items * @sizeOf(?TokenIndex) + expected_items * 3 * @sizeOf(NodeIndex);
+
+    var stack_fallback = std.heap.stackFallback(bytes_needed, p.comp.gpa);
+    const allocator = stack_fallback.get();
+
+    var names = std.ArrayList(?TokenIndex).initCapacity(allocator, expected_items) catch unreachable;
```

Comment:
I'm using `null` `TokenIndex` to represent a non-existent name but I think we could also get away with using `0` since a name will never be the 0th token.

Reply:
Doesn't matter now while it's a temporary value but it might be good if it ends up being saved in the tree.